vespalib/src/vespa/vespalib/stllike/cache.hpp

```diff
@@ -7,55 +7,224 @@
+    : SizeConstrainedLru(owner, capacity_bytes)
+{
+}
+
+template <typename P>
+cache<P>::ProbationarySegmentLru::~ProbationarySegmentLru() = default;
+
+template <typename P>
+void
+cache<P>::ProbationarySegmentLru::onRemove(const KeyT& k) {
+    SizeConstrainedLru::_owner.onRemove(k);
+}
+
+template <typename P>
+void
+cache<P>::ProbationarySegmentLru::onInsert(const KeyT& k) {
+    SizeConstrainedLru::_owner.onInsert(k);
+}
+
+template <typename P>
+cache<P>::ProtectedSegmentLru::ProtectedSegmentLru(cache& owner, size_t capacity_bytes)
+    : SizeConstrainedLru(owner, capacity_bytes)
+{
+}
+
+template <typename P>
+cache<P>::ProtectedSegmentLru::~ProtectedSegmentLru() = default;
+
+template <typename P>
+bool
+cache<P>::ProtectedSegmentLru::removeOldest(const typename P::value_type& kv) {
+    bool remove = SizeConstrainedLru::removeOldest(kv); // Updates own size if `remove == true`
+    if (remove) {
+        // Move back into probationary segment. This may shove the oldest entry out of it,
+        // which evicts it from the cache completely (no second chances for probationary elements).
+        // Note that we intercept this in removeOldest() and _not_ onRemove() since the latter
+        // only provides the _key_ of the element being removed, while removeOldest provides both
+        // the key and value. It is not obviously safe to resolve the key -> value from onRemove().
+        SizeConstrainedLru::_owner._probationary_segment.insert_and_update_size(kv.first, kv.second._value);
+    }
+    return remove;
+}
+
+template <typename P>
+cache<P>&
+cache<P>::maxElements(size_t probationary_elems, size_t protected_elems) {
+    std::lock_guard guard(_hashLock);
+    _probationary_segment.set_max_elements(probationary_elems);
+    _protected_segment.set_max_elements(protected_elems);
+    if (protected_elems == 0) { // If max protected elems == 0, this disables SLRU semantics
+        disable_slru();
+    }
+    return *this;
+}
+
+template <typename P>
+cache<P>&
 cache<P>::maxElements(size_t elems) {
-    Lru::maxElements(elems);
+    maxElements(elems, 0);
+    return *this;
+}
+
+template <typename P>
+cache<P>&
+cache<P>::setCapacityBytes(size_t probationary_sz, size_t protected_sz) {
+    std::lock_guard guard(_hashLock);
+    _probationary_segment.set_capacity_bytes(probationary_sz);
+    _protected_segment.set_capacity_bytes(protected_sz);
+    if (protected_sz == 0) { // If max protected size == 0, this disables SLRU semantics
+        disable_slru();
+    }
     return *this;
 }
 
-template< typename P >
-cache<P> &
+template <typename P>
+cache<P>&
 cache<P>::setCapacityBytes(size_t sz) {
-    _maxBytes.store(sz, std::memory_order_relaxed);
+    setCapacityBytes(sz, 0);
     return *this;
 }
 
-template< typename P >
+template <typename P>
+void
+cache<P>::disable_slru() {
+    _protected_segment.set_max_elements(0);
+    _protected_segment.set_capacity_bytes(0);
+    _protected_segment.evict_all();
```

Comment:
A swashbuckling tale of excitement has been added. tl;dr it's an edge case that is not expected in production so it just has to be correct, not particularly optimal. But if we add trimming functionality, this can be implicitly solved by setting capacities to zero and just trimming the segment instead (which will dump everything into the probationary segment by design).